src/nat.rs

```diff
@@ -32,8 +34,26 @@ pub fn get_public_ip_addr() -> Result<IpAddr, String> {
     }
 }
 
+pub fn udp_random_bind(start: u16, end: u16, tries: u32) -> Result<UdpSocket, String> {
+    let mut count = 0;
+    loop {
+        count += 1;
+
+        let rand_port = thread_rng().gen_range(start, end);
+        let addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), rand_port);
+
+        match UdpSocket::bind(addr) {
+            Result::Ok(val) => break Result::Ok(val),
+            Result::Err(_err) =>
+                if count >= tries  {
+                    break Result::Err("Failed to bind to port".to_string())
```

Comment:
this is a bit weird, you are dropping the OS error in favor of a local one.  i think i would rather return io::Result with the real error so we can debug.

So i would suggest to check `_err` for the expected error, which is an unavailable port, if its some other error, return that one immediately, otherwise return _err if `count == tries`

Reply:
Agreed. I also didn't like masking the error code. Pushing a patch to address this.